src/python/pants/backend/python/framework/stevedore/python_target_dependencies.py

```diff
@@ -0,0 +1,222 @@
+    PythonModuleOwnersRequest,
+)
+from pants.backend.python.framework.stevedore.target_types import (
+    AllStevedoreExtensionTargets,
+    StevedoreExtensionTargets,
+    StevedoreNamespace,
+    StevedoreNamespacesField,
+    StevedoreNamespacesProviderTargetsRequest,
+)
+from pants.backend.python.target_types import (
+    PythonDistribution,
+    PythonDistributionDependenciesField,
+    PythonDistributionEntryPointsField,
+    PythonTestsDependenciesField,
+    PythonTestsGeneratorTarget,
+    PythonTestTarget,
+    ResolvedPythonDistributionEntryPoints,
+    ResolvePythonDistributionEntryPointsRequest,
+)
+from pants.engine.addresses import Address, Addresses
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    AllTargets,
+    DependenciesRequest,
+    ExplicitlyProvidedDependencies,
+    FieldSet,
+    InferDependenciesRequest,
+    InferredDependencies,
+    Target,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import OrderedSet
+from pants.util.strutil import softwrap
+
+# -----------------------------------------------------------------------------------------------
+# Utility rules to analyze all StevedoreNamespace entry_points
+# -----------------------------------------------------------------------------------------------
+
+
+@rule(
+    desc=f"Find all `{PythonDistribution.alias}` targets with `{StevedoreNamespace.alias}` entry_points",
+    level=LogLevel.DEBUG,
+)
+def find_all_python_distributions_with_any_stevedore_entry_points(
+    targets: AllTargets,
+) -> AllStevedoreExtensionTargets:
+    return AllStevedoreExtensionTargets(
+        tgt
+        for tgt in targets
+        if tgt.has_field(PythonDistributionEntryPointsField)
+        and any(
+            # namespace aka category aka group
+            isinstance(namespace, StevedoreNamespace)
+            for namespace in (tgt[PythonDistributionEntryPointsField].value or {}).keys()
+        )
+    )
+
+
+@dataclass(frozen=True)
+class StevedoreExtensions:
+    """A mapping of stevedore namespaces to a list of targets that provide them.
+
+    Effectively, the targets are StevedoreExtension targets.
+    """
+
+    mapping: FrozenDict[str, tuple[Target, ...]]
+
+
+@rule(
+    desc=f"Create map of `{StevedoreNamespace.alias}` to `{PythonDistribution.alias}` targets",
+    level=LogLevel.DEBUG,
+)
+async def map_stevedore_extensions(
+    stevedore_extensions: AllStevedoreExtensionTargets,
+) -> StevedoreExtensions:
+    mapping: Mapping[str, list[Target]] = defaultdict(list)
+    for tgt in stevedore_extensions:
+        # namespace aka category aka group
+        for namespace, entry_points in (
+            tgt[PythonDistributionEntryPointsField].value or {}
+        ).items():
+            if isinstance(namespace, StevedoreNamespace):
+                mapping[str(namespace)].append(tgt)
```

Comment:
relates to previous comment, this could simply be `mapping[namespace].append(tgt)`